braindecode/eegneuralnet.py

```diff
@@ -114,3 +135,78 @@ def _default_callbacks(self):
             ),
             ("print_log", PrintLog()),
         ]
+
+    @abc.abstractmethod
+    def _get_n_outputs(self, y, classes):
+        pass
+
+    def _set_signal_args(self, X, y, classes):
+        # get kwargs from signal:
+        signal_kwargs = dict()
+        if isinstance(X, np.ndarray):
+            if y is None:
+                raise ValueError(f"y must be specified if X is a numpy array.")
+            self.log.info(f"Using numpy array to find signal-related parameters.")
+            Xshape = X.shape
+            signal_kwargs["n_times"] = Xshape[-1]
+            signal_kwargs["n_chans"] = Xshape[-2]
+            signal_kwargs['n_outputs'] = self._get_n_outputs(y, classes)
+
+        elif isinstance(X, torchDataset):
+            self.log.info(f"Using Dataset {X!r} to find signal-related parameters.")
+            X0 = X[0][0]
+            Xshape = X0.shape
+            signal_kwargs["n_times"] = Xshape[-1]
+            signal_kwargs["n_chans"] = Xshape[-2]
+            if (
+                    isinstance(X, BaseConcatDataset) and
+                    all(ds.targets_from == 'metadata' for ds in X.datasets)
+            ):
+                y_target = X.get_metadata().target
+                signal_kwargs['n_outputs'] = self._get_n_outputs(y_target, classes)
+            elif (
+                    isinstance(X, WindowsDataset) and
+                    X.targets_from == "metadata"
+            ):
+                y_target = X.windows.metadata.target
+                signal_kwargs['n_outputs'] = self._get_n_outputs(y_target, classes)
+        else:
+            raise ValueError(
+                f"X must be a numpy array or a Dataset, "
+                f"got {type(X)!r}."
+            )
+
+        # kick out missing kwargs:
+        module_kwargs = dict()
+        all_module_kwargs = inspect.signature(self.module.__init__).parameters.keys()
+        for k, v in signal_kwargs.items():
```

Comment:
why do we use only `signal_kwargs` here?

Reply:
if we start using kwargs for skorch we should use it for everything, otherwise we end up with reinitialization of model during fit to defualt parameters of the module